modules/core/src/main/java/org/apache/ignite/internal/util/typedef/X.java

```diff
@@ -547,24 +526,87 @@ public static boolean hasSuppressed(@Nullable Throwable t, @Nullable Class<? ext
         if (t == null || cls == null)
             return null;
 
-        for (Throwable th = t; th != null; th = th.getCause()) {
-            if (cls.isAssignableFrom(th.getClass()))
-                return (T)th;
+        @SuppressWarnings("unchecked")
+        T res = (T)searchForCause(t, null, cls);
 
-            for (Throwable n : th.getSuppressed()) {
-                T found = cause(n, cls);
+        return res;
+    }
 
-                if (found != null)
-                    return found;
-            }
+    /**
+     * Traverses `tree` of {@link Throwable} to find first cause/suppressed which meets test condition of predicate.
+     * Function is aware of possible circular references through tracking of tested objects.
+     *
+     * @param t Throwable - `root of the tree`.
+     * @param msg Possible throwable message.
+     * @param types Candidate types.
+     * @return First throwable meets test condition or {@code null} if none has matched.
+     */
+    @Nullable private static Throwable searchForCause(Throwable t, @Nullable String msg, Class<?>... types) {
+        Set<Throwable> dejaVu = Collections.newSetFromMap(new IdentityHashMap<>());
```

Comment:
Maybe use ThreadLocal WDYT? @ibessonov

Reply:
I don't think it is a good Idea. Almost the same gc pressure (we need to clean set before function start), but we get one level of indirection when using ThreadLocal. 

P.S. Same mechanism is used in jdk for printing throwable stacktrace. see java.lang.Throwable#printStackTrace(java.lang.Throwable.PrintStreamOrWriter)